### DISCUSSION TITLE
Reorganized page when refreshed

### DISCUSSION THREAD

**Discussion by Ethan-Mircheff** [community/community#147277] [Programming Help] [answered]
### Body

I am working on a [website](https://ethan-mircheff.github.io/pibillion/) that will show 1 billion digits of pi (I know its not practical but I thought it would be fun), and I'm having a bug where it will sometimes load with the strings of text switched places, so it ends up looking like this:
![Screenshot 2024-12-16 9 38 38 PM](https://github.com/user-attachments/assets/a3db5817-21bb-4582-9548-895bbcebc1d1)
instead of this:
![Screenshot 2024-12-16 9 38 44 PM](https://github.com/user-attachments/assets/e7ed1e87-ebc4-4a6c-a090-b3d58c0dacf4)
This only happens sometimes, and is fixed when I reload the page, so I can't figure out what the issue is. Right now this is my code:

```html
<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>One Billion Digits of Pi</title>
  <style>
  body {
    font-size: 20px;
    font-family: "Trebuchet MS", Veranda, sans-serif;
    display: inline-block;
    margin: 10px;
  }
    
  .digits {
    display: inline-block;
    font-family: serif;
  }
  </style>
  <script src="https://code.jquery.com/jquery-3.7.1.js"></script>
</head>
<body>

<h1>One Billion Digits of Pi:</h1>
  <p id="digits"></p>
  
<script>
  $(document).ready(function() {
  var paragraph = $("#digits");
  var loadFiles = ["digits.html","digits2.html"];
  $.each(loadFiles, function(index, file) {
    $.get(file, function(data) {

      var tempDiv = $("<div>").html(data);
      var textContent = tempDiv.text();
      var textNoSpaces = textContent.replace(/\s+/g, "");
      
      paragraph.append(textNoSpaces);
    });
  });
});
</script>
</body>
</html>
```

<br>
If you want to take a closer look, [this](https://github.com/Ethan-Mircheff/pibillion) is my repository with all the files. Thanks in advance for the help!

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by harmeetsingh11** [comment, marked as answer]
Hi @Ethan-Mircheff ,

The issue you're encountering is likely due to how the two files (`digits.html` and `digits2.html`) are being loaded asynchronously with the `$.get()` method. Since both requests happen at the same time and you append their results to the same `#digits` paragraph, it's possible that the data from the files is not being appended in the correct order or at the correct positions, leading to the text appearing "switched."

### Solution:

1. **Sequential Loading:** Instead of appending the content from both files simultaneously, load the second file only after the first one has finished loading. This way, the text is appended correctly.
You can achieve this by using `$.get()` within a callback function or using `async/await` with `fetch()`. Below are both ways.

## Using callback function

Here's the modified code using `$.get()` with callbacks to ensure sequential loading:

```html
<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>One Billion Digits of Pi</title>
  <style>
    body {
      font-size: 20px;
      font-family: "Trebuchet MS", Veranda, sans-serif;
      display: inline-block;
      margin: 10px;
    }
    
    .digits {
      display: inline-block;
      font-family: serif;
    }
  </style>
  <script src="https://code.jquery.com/jquery-3.7.1.js"></script>
</head>
<body>

<h1>One Billion Digits of Pi:</h1>
  <p id="digits"></p>
  
<script>
  $(document).ready(function() {
    var paragraph = $("#digits");
    var loadFiles = ["digits.html", "digits2.html"];
    
    function loadNextFile(index) {
      if (index < loadFiles.length) {
        $.get(loadFiles[index], function(data) {
          var tempDiv = $("<div>").html(data);
          var textContent = tempDiv.text();
          var textNoSpaces = textContent.replace(/\s+/g, "");
          paragraph.append(textNoSpaces);
          
          // Load the next file after the current one is processed
          loadNextFile(index + 1);
        });
      }
    }

    // Start loading from the first file
    loadNextFile(0);
  });
</script>
</body>
</html>
```

### Explanation of changes:
 - `loadNextFile()` is a recursive function that loads one file after the other, ensuring the correct order of the digits.
- We start by calling `loadNextFile(0)`, which begins with the first file (`digits.html`), then once that is loaded, the next file (`digits2.html`) is loaded.

## Using async/await and fetch() method

```html
<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>One Billion Digits of Pi</title>
  <style>
    body {
      font-size: 20px;
      font-family: "Trebuchet MS", Veranda, sans-serif;
      display: inline-block;
      margin: 10px;
    }
    
    .digits {
      display: inline-block;
      font-family: serif;
    }
  </style>
</head>
<body>

<h1>One Billion Digits of Pi:</h1>
  <p id="digits"></p>
  
<script>
  async function loadDigits() {
    const paragraph = document.getElementById("digits");
    const loadFiles = ["digits.html", "digits2.html"];

    // Load files sequentially
    for (let file of loadFiles) {
      try {
        const response = await fetch(file);
        const data = await response.text();

        // Remove spaces and append to the paragraph
        const textNoSpaces = data.replace(/\s+/g, "");
        paragraph.innerHTML += textNoSpaces;
      } catch (error) {
        console.error(`Error loading file ${file}:`, error);
      }
    }
  }

  // Start loading digits once the document is ready
  document.addEventListener("DOMContentLoaded", loadDigits);
</script>

</body>
</html>
```

### Explanation of changes:
1. `async function loadDigits()`: The function is declared `async` to allow the use of  `await` inside it, enabling us to wait for each file to load before moving to the next one.
2. `fetch(file)`: The `fetch()` API is used to load each file. It returns a promise that resolves with the response from the file.
3. `response.text()`: Once the response is obtained, we use `.text()`  to get the file's content as a string.
4. `replace(/\s+/g, "")`: This removes any spaces or newlines from the loaded content, ensuring the digits are concatenated without gaps.
5. `paragraph.innerHTML += textNoSpaces`: The cleaned-up content is appended to the #digits paragraph.
6. `Error handling`:  If any file fails to load, the error is logged, but the script continues to load the remaining files.

This should resolve the issue of the digits getting mixed up or switched, as the files will now load sequentially and be appended in the correct order.

Thanks

**Reply by Ethan-Mircheff** [reply]
Thanks so much for the help! Just to confirm, would I be able to add more files to this still? Because of the file size limits on GitHub, I will end up with 10-15 files for all of the digits. For example, would something like this work:

``` html
<script>
  async function loadDigits() {
    const paragraph = document.getElementById("digits");
    const loadFiles = ["digits.html", "digits2.html", "digits3.html", "digits4.html"];

    // Load files sequentially
    for (let file of loadFiles) {
      try {
        const response = await fetch(file);
        const data = await response.text();

        // Remove spaces and append to the paragraph
        const textNoSpaces = data.replace(/\s+/g, "");
        paragraph.innerHTML += textNoSpaces;
      } catch (error) {
        console.error(`Error loading file ${file}:`, error);
      }
    }
  }

  // Start loading digits once the document is ready
  document.addEventListener("DOMContentLoaded", loadDigits);
</script>

```

**Reply by harmeetsingh11** [reply]
Yes, your updated approach with additional files will work perfectly fine. The `for...of` loop in your `loadDigits` function will process any number of files listed in the `loadFiles` array sequentially. Here's a confirmation and a few clarifications:
- The `loadFiles` array can include as many files as you need (e.g., `digits3.html`, `digits4.html`, etc.).
- The function processes each file in the order they are listed, ensuring all the digits are appended in the correct sequence.
- There is no hard-coded limit to the number of files, so you can extend the array to include 10–15 files or more as necessary.

### Example:

```javascript
const loadFiles = [
  "digits.html",
  "digits2.html",
  "digits3.html",
  "digits4.html",
  "digits5.html",
  "digits6.html",
  "digits7.html",
  "digits8.html",
  "digits9.html",
  "digits10.html"
];
```

As long as the file names are correct and accessible, the script will load and append the digits sequentially.

### A Few Tips for Handling Many Files:

1. **Optimize File Access**:
   - Ensure all files are stored in the same directory or accessible path to avoid URL mismatches.
   - You might consider dynamically generating the file paths if they follow a pattern (e.g., `digits1.html`, `digits2.html`, etc.).

   ```javascript
   const loadFiles = Array.from({ length: 10 }, (_, i) => `digits${i + 1}.html`);
   ```

2. **Performance Consideration**:
   - While this solution works fine for 10–15 files, if you eventually scale up to hundreds of files, consider combining the files into larger chunks to minimize HTTP requests.

3. **Error Handling**:
   - If a file fails to load, you can display a message or skip it, but ensure the rest of the files still load properly.

   ```javascript
   console.warn(`Skipping file ${file} due to an error.`);
   ```

**Reply by Ethan-Mircheff** [reply]
Thanks for the tips! I am planning on having a total of 12-15 files, so that should work fairly well. I could definitely implement the dynamic file paths though, since they are following the pattern in your example, and they are all pretty similar. The error handling is also a good idea.
